tests/lowering/mod.rs

```diff
@@ -425,3 +421,138 @@ fn fundamental_multiple_type_parameters() {
         }
     }
 }
+
+#[test]
+fn tuples() {
+    lowering_success! {
+        program {
+            trait Foo { }
+
+            // `()` is an empty tuple
+            impl Foo for () { }
+            // `(i32,)` is a tuple
+            impl Foo for (i32,) { }
+            // `(i32)` is `i32` is a scalar
+            impl Foo for (i32) { }
+            impl Foo for (i32, u32) { }
+            impl Foo for (i32, u32, f32) { }
+        }
+    }
+}
+
+#[test]
+fn scalars() {
+    lowering_success! {
+        program {
+            trait Foo { }
+
+            impl Foo for i8 { }
+            impl Foo for i16 { }
+            impl Foo for i32 { }
+            impl Foo for i64 { }
+            impl Foo for i128 { }
+            impl Foo for isize { }
+            impl Foo for u8 { }
+            impl Foo for u16 { }
+            impl Foo for u32 { }
+            impl Foo for u64 { }
+            impl Foo for u128 { }
+            impl Foo for usize { }
+            impl Foo for f32 { }
+            impl Foo for f64 { }
+            impl Foo for bool { }
+            impl Foo for char { }
+        }
+    }
+
+    lowering_error! {
+        program {
+            struct i32 { }
+        }
+
+        error_msg {
+            "parse error: UnrecognizedToken { token: (8, Token(45, \"i32\"), 11), expected: [\"r#\\\"([A-Za-z]|_)([A-Za-z0-9]|_)*\\\"#\"] }"
+        }
+    }
+}
+
+#[test]
+fn tuple_trait_impl() {
+    let db = ChalkDatabase::with(
+        "
+trait Foo { }
+struct S1 { }
+impl Foo for (S1, S1) { }
```

Comment:
Are these tests trying to test the solver? If so, they should move to a file like tests/test/tuples.rs or something like that, and they use the `test!` macro like you see here:

https://github.com/rust-lang/chalk/blob/28cef6ff403d403e6ad2f3d27d944e9ffac1bce8/tests/test/auto_traits.rs#L5-L54